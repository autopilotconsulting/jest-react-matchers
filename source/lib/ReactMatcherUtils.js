import { propertiesForElement } from 'describe-react-element';
import flatten from 'array-flatten';

const ReactMatcherUtils = {
  elementsMatch: function(equals, actual, expected) {
    if (actual === null || typeof(actual) === 'undefined') { return false; }
    if (typeof(actual) === 'string') { return expected === actual; }
    if (expected.type !== actual.type) { return false; }

    const { children: expectedChildren, ...expectedProps } = propertiesForElement(expected);
    const { children: actualChildren, ...actualProps } = propertiesForElement(actual);

    const propertiesMatcher = jasmine.objectContaining(expectedProps);
    const propertiesMatch = equals(actualProps, propertiesMatcher);

    if (!propertiesMatch) { return false; }

    const expectedChildrenMatch = this.collectionHasMatchForAllElements(equals, actualChildren, expectedChildren);
    return expectedChildrenMatch;
  },

  collectionHasMatchForElement: function(equals, actualCollection, expected) {
    const test = (actual) => this.elementsMatch(equals, actual, expected);
    return actualCollection.some(test);
  },

  collectionHasMatchForAllElements: function(equals, actualCollection, expectedCollection) {
    if(typeof(expectedCollection) === 'undefined') { return true; }

    expectedCollection = flatten([expectedCollection]);
    actualCollection = flatten([actualCollection]);

    const test = (expected) => this.collectionHasMatchForElement(equals, actualCollection, expected);
    return expectedCollection.every(test);
  }
};

export default ReactMatcherUtils;