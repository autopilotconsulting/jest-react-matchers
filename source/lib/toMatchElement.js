import { describeReactElement } from 'describe-react-element';
import ReactMatcherUtils from './ReactMatcherUtils';

function toMatchElement(actualElement, expectedElement) {
  const result = { pass: ReactMatcherUtils.elementsMatch(this.equals, actualElement, expectedElement) };

  const actualDescription = describeReactElement(actualElement);
  const expectedDescription = describeReactElement(expectedElement);

  if (result.pass) {
    result.message = () => [
      'Expected element:',
      actualDescription,
      '',
      'not to match:',
      expectedDescription,
      '',
      'but it did!!',
    ].join("\n");
  } else {
    result.message = () => [
      'Expected element:',
      actualDescription,
      '',
      'To match:',
      expectedDescription
    ].join("\n");
  }

  return result;
}

export default toMatchElement;