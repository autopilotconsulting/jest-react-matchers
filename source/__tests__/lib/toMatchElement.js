import React from 'react';

import ReactMatcherUtils from '../../lib/ReactMatcherUtils.js';
import toMatchElement from '../../lib/toMatchElement.js';

expect.extend({ toMatchElement });

describe('ReactMatchers', () => {
  describe('#toMatchElement', () => {
    let matcherContext = { equals: () => { } };
    let mockElementsMatch;

    beforeEach(() => {
      mockElementsMatch = jest.spyOn(ReactMatcherUtils, 'elementsMatch');
    });

    afterEach(() => {
      mockElementsMatch.mockRestore();
    });

    it('should pass if elements match', () => {
      mockElementsMatch.mockImplementation(() => true);      

      const actual = <div />;
      const expected = <div />;

      expect(actual).toMatchElement(expected);
      expect(mockElementsMatch).toHaveBeenCalledWith(expect.anything(), actual, expected);
    });

    it('should be negatable', () => {
      mockElementsMatch.mockImplementation(() => false)

      const actual = <div />;
      const expected = <span />;

      expect(actual).not.toMatchElement(expected);
      expect(mockElementsMatch).toHaveBeenCalledWith(expect.anything(), actual, expected);
    });

    it('should describe a failed match', () => {
      mockElementsMatch.mockImplementation(() => false)
      const result = toMatchElement.call(matcherContext, <div />, <span />);
      
      expect(result.message).toMatchSnapshot();
    });

    it('should describe a failed negated match', () => {
      mockElementsMatch.mockImplementation(() => true)
      const result = toMatchElement.call(matcherContext, <div />, <div />);
      
      expect(result.message).toMatchSnapshot();
    });
  });
});