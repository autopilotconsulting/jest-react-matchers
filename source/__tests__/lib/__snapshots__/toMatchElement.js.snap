// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactMatchers #toMatchElement should describe a failed match 1`] = `
"Expected element:
<div />

To match:
<span />"
`;

exports[`ReactMatchers #toMatchElement should describe a failed negated match 1`] = `
"Expected element:
<div />

not to match:
<div />

but it did!!"
`;
